test/e2e/framework/firewall_util.go

```diff
@@ -302,6 +302,72 @@ func PackProtocolsPortsFromFirewall(alloweds []*compute.FirewallAllowed) []strin
 	return protocolPorts
 }
 
+type portRange struct {
+	protocol  string
+	startPort int64
+	endPort   int64
+}
+
+// isPortsSubset asserts that the "requiredPorts" are covered by the "coverage" ports.
+// requiredPorts - must be single-port, examples: 'tcp/50', 'udp/80'.
+// coverage - single or port-range values, example: 'tcp/50', 'udp/80-1000'.
+// Returns true if every requiredPort exists in the list of coverage rules.
+func isPortsSubset(requiredPorts, coverage []string) error {
+	for _, reqPort := range requiredPorts {
+		r := strings.Split(reqPort, "/")
+		port, err := strconv.Atoi(r[1])
+		if err != nil {
+			return err
+		}
+		var covered bool
+
+	coveredPortLoop:
+		for _, c := range coverage {
+			// Assert IPProtocol value is identical.
+			cFields := strings.Split(c, "/")
+			if r[0] != cFields[0] {
+				continue
+			}
+
+			if len(cFields) != 2 {
+				return fmt.Errorf("unexpected allowed string %q", c)
+			}
+
+			covPortRange := strings.Split(cFields[1], "-")
+			switch len(covPortRange) {
+			case 1:
+				// Check the ports are equal.
+				if r[1] == covPortRange[0] {
+					covered = true
+					break coveredPortLoop
+				}
+			case 2:
+				// Check the port exists in the covered port's range.
+				start, err := strconv.Atoi(covPortRange[0])
+				if err != nil {
+					return err
+				}
+				end, err := strconv.Atoi(covPortRange[1])
+				if err != nil {
+					return err
+				}
+
+				if port >= start && port <= end {
+					covered = true
+					break coveredPortLoop
+				}
+			default:
+				return fmt.Errorf("unexpected range value %q", cFields[1])
+			}
+		}
+
+		if !covered {
+			return fmt.Errorf("%q is not covered", reqPort)
```

Comment:
Maybe helpful to also include `coverage` in the error?

Reply:
Done. thanks